pymcspearman/pymckendall.py

```diff
@@ -0,0 +1,128 @@
+"""
+pymckendall.py
+
+Python implementation of Isobe, Feigelson & Nelson (1986) method for calculating Kendall
+tau rank correlation coefficient with uncertainties on censored data (upper/lowerlimit).
+
+"""
+
+import numpy as np
+from matplotlib import pyplot as plt
+import pandas as pd
+import scipy.stats as st
+
+#generalized kendall tau test described in Isobe,Feigelson & Nelson 1986
+#need to come up some way to vectorize this function, very slow
+def kendall(x,y,xlim,ylim):
+    #x, y are two arrays, either may contain censored data
+    #xlim, ylim are arrays indicating if x or y are lower or upperlimits,-1--lowerlimit,+1--upperlimit,0--detection
+    num=len(x)#x,y should have same length
+    #set up pair counters
+    a=np.zeros((num,num))
+    b=np.zeros((num,num))
+    
+    for i in range(num):
+        for j in range(num):
+            if x[i]==x[j]:
+                a[i,j]=0
+            elif x[i] > x[j]: #if x[i] is definitely > x[j]
+                if (xlim[i]==0 or xlim[i]==-1) and (xlim[j]==0 or xlim[j]==1):
+                    a[i,j]=-1
+            else: #if x[i] is definitely < x[j], all other uncertain cases have aij=0
+                if (xlim[i]==0 or xlim[i]==1) and (xlim[j]==0 or xlim[j]==-1):
+                    a[i,j]=1
+            
+    for i in range(num):
+        for j in range(num):
+            if y[i]==y[j]:
+                b[i,j]=0
+            elif y[i] > y[j]:
+                if (ylim[i]==0 or ylim[i]==-1) and (ylim[j]==0 or ylim[j]==1):
+                    b[i,j]=-1
+            else:
+                 if (ylim[i]==0 or ylim[i]==0) and (ylim[j]==0 or ylim[j]==-1):
+                    b[i,j]=1
+                    
+            
+    S = np.sum(a*b)
+    var = (4/(num*(num-1)*(num-2)))*(np.sum(a*np.sum(a,axis=1,keepdims=True))-np.sum(a*a))\
+    *(np.sum(b*np.sum(b,axis=1,keepdims=True))-np.sum(b*b))+(2/(num*(num-1)))*np.sum(a*a)*np.sum(b*b)
+    z=S/np.sqrt(var)
+    tau=z*np.sqrt(2*(2*num+5))/(3*np.sqrt(num*(num-1)))
+    pval=st.norm.sf(abs(z))*2
+    return tau,pval
+
+
+def pymckendall(x, y, xlim, ylim, dx=None, dy=None, Nboot=10000, Nperturb=10000,
+    bootstrap=True,
+    perturb=True,
+    percentiles=(16,50,84), return_dist=False):
+    """
+    Compute spearman rank coefficient with uncertainties using several methods.
```

Comment:
minor: this should be changed to say Kendall tau

Reply:
changed